r-package/grf/R/input_utilities.R

```diff
@@ -197,12 +199,19 @@ observation_weights <- function(forest) {
     forest$sample.weights * length(forest$Y.orig) / sum(forest$sample.weights)
   }
   if (length(forest$clusters) == 0) {
+  observation.weight <- sample.weights
+  }
+  if (length(forest$clusters) != 0 && is.null(forest$sample.weights)) {
     observation.weight <- sample.weights
+    if (forest$clusters.subsample) {
+      clust.factor <- factor(forest$clusters)
+      inverse.counts <- 1 / as.numeric(Matrix::colSums(Matrix::sparse.model.matrix(~ clust.factor + 0)))
+      observation.weight <- sample.weights * inverse.counts[as.numeric(clust.factor)]
+    }
   } else {
-    clust.factor <- factor(forest$clusters)
-    inverse.counts <- 1 / as.numeric(Matrix::colSums(Matrix::sparse.model.matrix(~ clust.factor + 0)))
-    observation.weight <- sample.weights * inverse.counts[as.numeric(clust.factor)]
+    observation.weight <- sample.weights
   }
+
```

Comment:
For clarity, can we write this function something like this? Like that, we clearly disambiguate between all possible cases of clustering vs sample weighting

if no sample weights provided
&nbsp;  if no clusters provided or we don't do unit per-cluster  weighting
&nbsp;&nbsp;     raw weights are all ones
 &nbsp; if we do unit per-cluster weighting
 &nbsp;&nbsp;    raw weights is what's on 207-209 above
if no sample weights provided
&nbsp;  if no clusters provided or we don't do unit per-cluster  weighting
  &nbsp;&nbsp;   raw weights are sample weights
&nbsp;  if we do unit per-cluster weighting
 &nbsp;&nbsp;    ERROR

return raw weights / sum raw weights

Reply:
Do you mean return just `raw weights`? If `raw weights / sum raw weights` some existing tests with no clustering or sample weights will fail?

Reply:
Which tests fail? I didn't thing normalizing or not normalizing the weights mattered, but thought it'd be safer to return normalized weights just in case (always good to be strict on outputs).

Reply:
10 tests fails: https://travis-ci.com/grf-labs/grf/jobs/259829947#L1564